### ISSUE TITLE
PP microbatching + block-causal FlexAttention composability issue

### ISSUE THREAD

**Issue by tianyu-l** [pytorch/torchtitan#1723]
Today `init_attention_mask` is called before PP does microbatch split
https://github.com/pytorch/torchtitan/blob/main/torchtitan/train.py#L421

I haven't tested, but this will likely cause wrong block mask being applied to non-first microbatches.
E.g. consider a local batch
```
batch = [
  b0,
  b1,
  b2,
  b3,
]
```
Block mask `mask` is created for `batch`, but after (say, size-1) microbatching `mask` will be used for 4 different smaller batches `[b0]`, `[b1]`, `[b2]`, `[b3]`. For `[b0]` it might be fine, but for the others the mask is wrong.

The solution could be either of
1. do `init_attention_mask` after microbatching.
2. when PP does microbatching, modify the block mask as well.

cc @fegin @H-Huang @drisspg

**Comment by fegin** [comment]
~~3. If we are going to make CP register a model forward hook, the hook can modify the BlockMask, provided that CP knows PP's information.~~

The above only work when CP is enabled but this issue can happen even if CP is not enabled.

**Comment by ezyang** [comment]
Just a question for my curiosity: the way that init_attention_mask works seems a bit... ugly to me? https://github.com/pytorch/torchtitan/blob/60645bc78cf4138ae54e8f29a8b24bf0ccd64404/torchtitan/models/attention.py#L186 Is the reason the block masks member is a class variable because you want to share them across all attention blocks? The general pattern smacks of wanting some sort of laziness, where you init it the first time you get a batch for real (so at that point you get a microbatch) (although this is not easily compilable!)

**Comment by drisspg** [comment]
The pattern that we have recommended is to have blockmask be an input to your model. Torchtune had set this up and made a thin type that could either be a blockmask or tensor or none so that it could be used for sdpa or flex

**Comment by ezyang** [comment]
Yes, and then it would be obvious that you also have to microbatch it when doing PP

**Comment by tianyu-l** [comment]
One concern we had in the past is that
Different layers of TransformerBlock have different attention masks, e.g. some layers use sliding window, some layers use full causal, etc. So it will become a bit cumbersome to pass all these masks around and let PP microbatch everything.

Another option is to simply pass the `eos_id` around (which is the only thing affecting the masks from iteration to iteration), which PP microbatching doesn't need to do anything.

This is further complicated by CP inputs sharding with load balancing. See discussions in https://github.com/pytorch/torchtitan/issues/1728#issuecomment-3314783153

**Comment by fegin** [comment]
> The pattern that we have recommended is to have blockmask be an input to your model. Torchtune had set this up and made a thin type that could either be a blockmask or tensor or none so that it could be used for sdpa or flex

We are thinking to do this as well. But as @tianyu-l described, the main concern was different layers masks and how does it compose 

But PP cannot  micro-batch BlockMask atm as it doesn't understand BlockMask. I'm wondering if we should let PP understands BlockMask, just like PP understands torch.Tensor as BlockMask is from torch core. cc., @H-Huang

**Comment by H-Huang** [comment]
In PP you do `schedule.step(input_batch, targets=targets)`, targets (`targets.shape == input_batch.shape`) are split into microbatches by PP , but are not communicated through pipeline stages. They are only used by the last stage computing the loss function by indexing based on microbatch (`targets[microbatch_index]`).

Can we do something similar for BlockMask? If you pass in a list of BlockMasks, one for each microbatch, then each stage's forward can use the BlockMask intended for that microbatch?

**Comment by ezyang** [comment]
A compromise here is to use some sort of dynamically scoped variable, which you fetch the masks out of. The important thing is that they are still inputs at the top level so PP knows about them, but at least you don't have to plumb them through all the function calls.

**Comment by awaelchli** [comment]
The way I've worked around this is by moving the block mask construction into the model forward itself. Then PP micro-batching operates on the batch and doc ids first, and I can construct the block mask for the micro-batches individually. It did not have a measurable perf impact. 

Being able to micro-batch the BlockMask would be useful imo. Perhaps giving the user an option to configure how certain inputs get split could also be considered.

**Comment by fegin** [comment]
Okay, I draft https://github.com/pytorch/pytorch/pull/164111/, which can split BlockMask. cc., @drisspg @H-Huang
